relay/dataplane/s3_client.go

```diff
@@ -0,0 +1,241 @@
+package dataplane
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"sync"
+	"time"
+)
+
+var _ S3Client = &s3Client{}
+
+type s3Client struct {
+	// config is the configuration for the client.
+	config *S3Config
+	// ctx is the context for the client.
+	ctx context.Context
+	// cancel is called to cancel the context.
+	cancel context.CancelFunc
+	// the AWS S3 handle to use to talk to S3.
+	svc *s3.S3
+	// tasks are placed into this channel to be processed by workers.
+	tasks chan func()
+	// this wait group is completed when all workers have finished.
+	wg *sync.WaitGroup
```

Comment:
I think what you're doing here can be simplified & abstracted out by using `workerpool` (ex. https://github.com/Layr-Labs/eigenda/blob/master/api/clients/retrieval_client.go#L181)

Reply:
This is a neat little library, good to know about. Done.